### PR TITLE
Decode content-addressed gem metadata in remote CLI output

### PR DESCRIPTION
**Pull request by girachawda**
## What

Addresses: https://github.com/ruby/rubygems/issues/9729

Update remote RubyGems CLI output so content-addressable / skinny gems are displayed using real gem metadata instead of the content-address hash.

This PR covers:

- `gem list -r`
- `gem search -r`
- `gem info -r`

For compact `gem list` / `gem search` output, skinny metadata is formatted with these readability rules:

- different gem versions are shown on separate lines
- different platforms are shown on separate lines
- multiple Ruby ABIs for the same version + platform stay on the same line
- when Ruby ABI metadata is shown, both `Platform:` and `Ruby ABI:` are labeled
<details>
<summary>Example output shapes:</summary>

Same version + same platform + one Ruby ABI:

```text
a (1 Platform: x86_64-linux, Ruby ABI: 3.3)
```

Same version + same platform + multiple Ruby ABIs:

```text
a (1 Platform: x86_64-linux, Ruby ABI: 3.3, 3.4)
```

Same version + multiple platforms:

```text
a (1 Platform: x86_64-linux, Ruby ABI: 3.3
   1 Platform: x86_64-linux-musl, Ruby ABI: 3.4)
```

Multiple versions:

```text
a (3 Platform: arm64-darwin, Ruby ABI: 3.4
   2 Platform: x86_64-linux, Ruby ABI: 3.4
   1 Platform: x86_64-linux, Ruby ABI: 3.3)
```

Multiple versions + multiple platforms + multiple Ruby ABIs:

```text
a (2 Platform: arm64-darwin, Ruby ABI: 3.5
   2 Platform: x86_64-linux, Ruby ABI: 3.4
   1 Platform: arm64-darwin, Ruby ABI: 3.4
   1 Platform: x86_64-linux, Ruby ABI: 3.3, 3.4)
```

Fat/platform binary without Ruby ABI metadata:

```text
a (1 x86_64-linux)
```
</details>

## Why

Skinny gems allow multiple binary gems to exist for the same gem name, version, and platform, split by Ruby ABI.

For example:

```text
my_gem 1.0.0 x86_64-linux Ruby ABI: 3.3
my_gem 1.0.0 x86_64-linux Ruby ABI: 3.4
```

Those variants cannot both use the traditional filename suffix:

```text
my_gem-1.0.0-x86_64-linux.gem
```

So content-addressable gems use a hash/token in the version/platform slot:

```text
my_gem-1.0.0-abcdef12.gem
my_gem-1.0.0-fedcba98.gem
```

That hash is useful internally, but it is not meaningful to users. Users need to know which platform and Ruby ABI a gem supports, not its content-address token.

This PR makes the remote query CLI commands decode that metadata and display the useful information.

## Tophat

<details>
<summary>Manual tophat setup and instructions</summary>

Run this from the repo root. It starts a tiny local compact-index server and checks the main skinny/CA display shapes:

- same version + same platform + multiple Ruby ABIs stays on one line
- same version + multiple platforms splits by platform
- multiple versions split by version
- multiple versions + multiple platforms + multiple Ruby ABIs keeps Ruby ABI lists grouped per version/platform
- fat/platform binaries with multiple platforms show the version once
- mixed content-addressable and fat/platform versions display together correctly
- `gem info -r` groups Ruby ABI details under `Platforms:`

```sh
TMPDIR_CA=$(mktemp -d)
GEMHOME_CA=$(mktemp -d)
PORT_CA=$(ruby -rsocket -e 's = TCPServer.new("127.0.0.1", 0); print s.addr[1]; s.close')
MARSHAL_VERSION=$(ruby --disable-gems -Ilib -e 'require "rubygems"; print Gem.marshal_version')
SERVER_PID=""
trap 'if [ -n "${SERVER_PID}" ] && kill -0 "${SERVER_PID}" 2>/dev/null; then kill "${SERVER_PID}" 2>/dev/null || true; wait "${SERVER_PID}" 2>/dev/null || true; fi; rm -rf "$TMPDIR_CA" "$GEMHOME_CA"' EXIT

mkdir -p "$TMPDIR_CA/info" "$TMPDIR_CA/quick/Marshal.$MARSHAL_VERSION"

cat > "$TMPDIR_CA/versions" <<'VERSIONS'
created_at: 2026-01-01T00:00:00Z
---
a 1-abcdef12,1-fedcba98 0000
b 1-abcdef12,1-fedcba98 0000
c 1-abcdef12,2-fedcba98,3-12345678 0000
d 1-aaaa1111,1-bbbb2222,1-cccc3333,2-dddd4444,2-eeee5555 0000
e 1-x86_64-linux,1-arm64-darwin 0000
m 1-x86_64-linux,2-abcdef12,3-fedcba98 0000
VERSIONS

cat > "$TMPDIR_CA/info/a" <<'INFO'
---
1-abcdef12 |checksum:123,ruby:~> 3.3.0,platform:= x86_64-linux
1-fedcba98 |checksum:456,ruby:~> 3.4.0,platform:= x86_64-linux
INFO

cat > "$TMPDIR_CA/info/b" <<'INFO'
---
1-abcdef12 |checksum:123,ruby:~> 3.3.0,platform:= x86_64-linux
1-fedcba98 |checksum:456,ruby:~> 3.4.0,platform:= x86_64-linux-musl
INFO

cat > "$TMPDIR_CA/info/c" <<'INFO'
---
1-abcdef12 |checksum:123,ruby:~> 3.3.0,platform:= x86_64-linux
2-fedcba98 |checksum:456,ruby:~> 3.4.0,platform:= x86_64-linux
3-12345678 |checksum:789,ruby:~> 3.4.0,platform:= arm64-darwin
INFO

cat > "$TMPDIR_CA/info/d" <<'INFO'
---
1-aaaa1111 |checksum:111,ruby:~> 3.3.0,platform:= x86_64-linux
1-bbbb2222 |checksum:222,ruby:~> 3.4.0,platform:= x86_64-linux
1-cccc3333 |checksum:333,ruby:~> 3.4.0,platform:= arm64-darwin
2-dddd4444 |checksum:444,ruby:~> 3.4.0,platform:= x86_64-linux
2-eeee5555 |checksum:555,ruby:~> 3.5.0,platform:= arm64-darwin
INFO

cat > "$TMPDIR_CA/info/m" <<'INFO'
---
2-abcdef12 |checksum:123,ruby:~> 3.3.0,platform:= x86_64-linux
3-fedcba98 |checksum:456,ruby:~> 3.4.0,platform:= arm64-darwin
INFO

ruby --disable-gems -Ilib -rzlib -e 'require "rubygems"; address = ARGV.fetch(0); abi = ARGV.fetch(1); spec = Gem::Specification.new("a", "1"); spec.platform = "x86_64-linux"; spec.content_address = address; spec.summary = "manual tophat summary for ruby #{abi}"; spec.homepage = "http://example.com"; spec.authors = ["Tophat User"]; print Zlib::Deflate.deflate(Marshal.dump(spec))' abcdef12 3.3 > "$TMPDIR_CA/quick/Marshal.$MARSHAL_VERSION/a-1-abcdef12.gemspec.rz"

ruby --disable-gems -Ilib -rzlib -e 'require "rubygems"; address = ARGV.fetch(0); abi = ARGV.fetch(1); spec = Gem::Specification.new("a", "1"); spec.platform = "x86_64-linux"; spec.content_address = address; spec.summary = "manual tophat summary for ruby #{abi}"; spec.homepage = "http://example.com"; spec.authors = ["Tophat User"]; print Zlib::Deflate.deflate(Marshal.dump(spec))' fedcba98 3.4 > "$TMPDIR_CA/quick/Marshal.$MARSHAL_VERSION/a-1-fedcba98.gemspec.rz"

python3 -m http.server "$PORT_CA" --bind 127.0.0.1 --directory "$TMPDIR_CA" >/tmp/rubygems-ca-tophat.log 2>&1 &
SERVER_PID=$!
sleep 0.5

SOURCE_CA="http://127.0.0.1:$PORT_CA"
export GEM_HOME="$GEMHOME_CA"
export GEM_PATH="$GEMHOME_CA"

echo "== same version + same platform + multiple Ruby ABIs =="
ruby --disable-gems -Ilib exe/gem list a -r --clear-sources --source "$SOURCE_CA"
ruby --disable-gems -Ilib exe/gem search a -r --clear-sources --source "$SOURCE_CA"
ruby --disable-gems -Ilib exe/gem info a -r --clear-sources --source "$SOURCE_CA"

echo
echo "== same version + multiple platforms =="
ruby --disable-gems -Ilib exe/gem list b -r --clear-sources --source "$SOURCE_CA"

echo
echo "== multiple versions =="
ruby --disable-gems -Ilib exe/gem list c -r --all --clear-sources --source "$SOURCE_CA"

echo
echo "== multiple versions + multiple platforms + multiple Ruby ABIs =="
ruby --disable-gems -Ilib exe/gem list d -r --all --clear-sources --source "$SOURCE_CA"

echo
echo "== fat/platform binary with multiple platforms =="
ruby --disable-gems -Ilib exe/gem list e -r --clear-sources --source "$SOURCE_CA"

echo
echo "== mixed content-addressable and fat/platform versions =="
ruby --disable-gems -Ilib exe/gem list m -r --all --clear-sources --source "$SOURCE_CA"
ruby --disable-gems -Ilib exe/gem search m -r --all --clear-sources --source "$SOURCE_CA"
```

</details>

<details>
<summary>Manual tophat output</summary>

```text
== same version + same platform + multiple Ruby ABIs ==
a (1 Platform: x86_64-linux, Ruby ABI: 3.3, 3.4)
a (1 Platform: x86_64-linux, Ruby ABI: 3.3, 3.4)
a (1)
    Platforms:
        x86_64-linux Ruby ABI: 3.3, 3.4
    Author: Tophat User
    Homepage: http://example.com

    manual tophat summary for ruby 3.3

== same version + multiple platforms ==
b (1 Platform: x86_64-linux, Ruby ABI: 3.3
   1 Platform: x86_64-linux-musl, Ruby ABI: 3.4)

== multiple versions ==
c (3 Platform: arm64-darwin, Ruby ABI: 3.4
   2 Platform: x86_64-linux, Ruby ABI: 3.4
   1 Platform: x86_64-linux, Ruby ABI: 3.3)

== multiple versions + multiple platforms + multiple Ruby ABIs ==
d (2 Platform: arm64-darwin, Ruby ABI: 3.5
   2 Platform: x86_64-linux, Ruby ABI: 3.4
   1 Platform: arm64-darwin, Ruby ABI: 3.4
   1 Platform: x86_64-linux, Ruby ABI: 3.3, 3.4)

== fat/platform binary with multiple platforms ==
e (1 arm64-darwin x86_64-linux)
== mixed content-addressable and fat/platform versions ==
m (3 Platform: arm64-darwin, Ruby ABI: 3.4
   2 Platform: x86_64-linux, Ruby ABI: 3.3
   1 Platform: x86_64-linux)
m (3 Platform: arm64-darwin, Ruby ABI: 3.4
   2 Platform: x86_64-linux, Ruby ABI: 3.3
   1 Platform: x86_64-linux)
```

The content-address hashes (`abcdef12`, `fedcba98`, etc.) should not be displayed in CLI output. For compact `list` / `search` output, different versions and different platforms are split onto separate lines, while multiple `Ruby ABI:` values for the same version/platform stay grouped on one line.

</details>

### CODE REVIEW

File: lib/rubygems/query_utils.rb [hunk 1]
```diff
@@ -342,6 +359,28 @@ def spec_platforms(entry, platforms)
     end
   end
 
+  def spec_ruby_abis(entry, ruby_abis)
```
Comment (girachawda) [inline lib/rubygems/query_utils.rb]: Prints detailed Ruby ABI info like `Ruby ABI: 3.3` or `Ruby ABIs: 3.3, 3.4`                                                                                                                              
 or grouped by version if needed:        
```
Ruby ABIs:                                                                                                                       
       1.0.0: 3.3                                                                                                                   
       2.0.0: 3.4         
```

File: lib/rubygems/query_utils.rb [hunk 2]
```diff
@@ -266,7 +271,18 @@ def entry_versions(entry, name_tuples, platforms, specs)
     entry << " (#{list.join ", "})"
   end
 
-  def make_entry(entry_tuples, platforms)
+  def version_platform_labels(version, platforms, platform_ruby_abis)
```
Comment (girachawda) [inline lib/rubygems/query_utils.rb]: Builds labels like `x86_64-linux ruby-abi:3.3 ruby-abi:3.4`

File: lib/rubygems/safe_marshal.rb [hunk 3]
```diff
@@ -51,7 +51,7 @@ module SafeMarshal
         @name @requirement @prerelease @version_requirement @version_requirements @type
         @force_ruby_platform
       ],
-      "Gem::NameTuple" => %w[@name @version @platform],
+      "Gem::NameTuple" => %w[@name @version @platform @content_address @ruby_abi],
```
Comment (girachawda) [inline lib/rubygems/safe_marshal.rb:54]: Compatibility-wise to see if adding these these two fields affects things:

- **Backwards compatibility:** new RubyGems can still read old marshalled `NameTuple`s because old payloads simply do not include `@content_address` or `@ruby_abi`. In that case, the new accessors return `nil`.

- **Forwards compatibility for normal gems and fat binaries:** we avoid setting `@content_address` and `@ruby_abi` when they are `nil`, so normal/fat `NameTuple`s still marshal with the old shape: `@name`, `@version`, `@platform`. Fat binaries also keep using the normal platform path, so older clients are not exposed to extra CA-only metadata.

- **Skinny CA gems:** tuples that actually have `@content_address` / `@ruby_abi` require a CA-aware RubyGems client. Older clients should be kept away from those CA entries through the compact-index compatibility rules.

File: lib/rubygems/name_tuple.rb [hunk 4]
```diff
@@ -46,11 +48,15 @@ def self.null
...
+      "#{@name}-#{@version}-#{@content_address}"
     else
-      "#{@name}-#{@version}-#{@platform}"
+      case @platform
+      when nil, "", Gem::Platform::RUBY
+        "#{@name}-#{@version}"
+      else
+        "#{@name}-#{@version}-#{@platform}"
+      end
     end
```
Comment (jenshenny) [inline lib/rubygems/name_tuple.rb:57]: I don't fully understand this conditional, there's a case statement that has a if statement with a nested conditional. Can we simplify this (I think the Specification class has a similar impl of this method)?

File: lib/rubygems/name_tuple.rb [hunk 5]
```diff
@@ -84,18 +90,27 @@ def to_a
...
 
   alias_method :to_s, :inspect # :nodoc:
 
   def <=>(other)
-    [@name, @version, Gem::Platform.sort_priority(@platform)] <=>
-      [other.name, other.version, Gem::Platform.sort_priority(other.platform)]
+    sort_key <=> other.sort_key
+  end
+
+  def sort_key # :nodoc:
```
Comment (jenshenny) [inline lib/rubygems/name_tuple.rb:109]: Could this method be private?

**Review by jenshenny — COMMENTED** [review]
I have some question swith how we are outputting these results!

```
a (1 x86_64-linux ruby-abi:3.3 ruby-abi:3.4)
a (1 x86_64-linux ruby-abi:3.3 ruby-abi:3.4)
a (1)
    Platform: x86_64-linux
    Ruby ABIs: 3.3, 3.4
    Author: Tophat User
    Homepage: http://example.com

    manual tophat summary for ruby 3.4
```
Could we condense `ruby-abi:3.3 ruby-abi:3.4` to `ruby-abi: 3.3, 3.4`? Do these commands return multiple versions with multiple platforms and multiple ABIs? What would the output look like?

For example, could something like this happen:

```
a (1)
    Platform: x86_64-linux, x86_64-linux-musl
    Ruby ABIs: 3.3, 3.4
    Author: Tophat User
    Homepage: http://example.com
```

We could lose some info if one platform differed in the ABIs it supports.

File: lib/rubygems/name_tuple.rb (hunk 5 above)
Comment (girachawda) [reply lib/rubygems/name_tuple.rb:109]: It can’t be private as written because `<=>` calls `other.sort_key`. I can make it protected instead, which keeps the helper hidden from public API but still usable for comparing two NameTuples.

**Comment by girachawda** [comment]
@jenshenny -> changed it so that this is how its outputted for multiple platforms! good catch! i also added a test to catch this case as well.

- same platform with multiple Ruby ABIs:
  ```text
  a (1 x86_64-linux ruby-abi: 3.3, 3.4)
  ```

- multiple platforms with different Ruby ABIs:
  ```text
  b (1 x86_64-linux ruby-abi: 3.3, x86_64-linux-musl ruby-abi: 3.4)
  ```

File: lib/rubygems/name_tuple.rb (hunk 4 above)
Comment (girachawda) [reply lib/rubygems/name_tuple.rb:57]: Yep, I simplified this. I split off the methods to add more readability. So the full name gets built once. I tried to follow the logic behaviour in Gem::BasicSpecification#full_name.

File: lib/rubygems/name_tuple.rb [hunk 6]
```diff
@@ -6,16 +6,18 @@
...
-  def initialize(name, version, platform = Gem::Platform::RUBY)
+  def initialize(name, version, platform = Gem::Platform::RUBY, content_address: nil, ruby_abi: nil)
     @name = name
     @version = version
 
     platform &&= platform.to_s
     platform = Gem::Platform::RUBY if !platform || platform.empty?
     @platform = platform
+    @content_address = content_address unless content_address.nil?
+    @ruby_abi = ruby_abi unless ruby_abi.nil?
```
Comment (jenshenny) [inline lib/rubygems/name_tuple.rb:17]: The guards are redundant since `@content_address` is `nil` by default.

```suggestion
    @content_address = content_address
    @ruby_abi = ruby_abi
```

File: lib/rubygems/query_utils.rb [hunk 7]
```diff
@@ -200,9 +200,15 @@ def output_versions(output, versions)
...
       platforms = Hash.new {|h,version| h[version] = [] }
+      ruby_abis = Hash.new {|h,version| h[version] = [] }
+      platform_ruby_abis = Hash.new {|h,version| h[version] = Hash.new {|hh,platform| hh[platform] = [] } }
 
       matching_tuples.each do |n, _|
         platforms[n.version] << n.platform if n.platform
+        next unless n.ruby_abi
+
+        ruby_abis[n.version] << n.ruby_abi
+        platform_ruby_abis[n.version][n.platform] << n.ruby_abi
```
Comment (jenshenny) [inline lib/rubygems/query_utils.rb]: Do we need both `ruby_abis` and `platform_ruby_abis`? To my knowledge we always group ABIs by platform.

File: lib/rubygems/query_utils.rb
Comment (jenshenny) [inline lib/rubygems/query_utils.rb:1]: The logic for the formatting is hard to follow (lots of branches and checks). I don't have any concrete suggestions to give atm but can you spend some time to figure out if there's a more readable way to write this logic?

File: lib/rubygems/source.rb [hunk 8]
```diff
@@ -259,13 +259,36 @@ def load_compact_index_specs(type)
...
         next if version.prerelease? != (type == :prerelease)
 
-        Gem::NameTuple.new(name, version, platform || "ruby")
+        suffix ||= "ruby"
+        platform = suffix
+        content_address = nil
+        ruby_abi = nil
+
+        if Gem::ContentAddress.match?(suffix)
+          info_rows ||= compact_index_info_rows(name)
```
Comment (jenshenny) [inline lib/rubygems/source.rb]: We're requesting the info endpoint for every CA gem. I think we should only call `compact_index_info` when we really need it the platform / Ruby ABI info.

**Review by jenshenny — COMMENTED** [review]
It's getting to a good place! I think the formatting is much more readable.

File: lib/rubygems/name_tuple.rb [hunk 9]
```diff
@@ -120,6 +136,6 @@ def ==(other)
   alias_method :eql?, :==
 
   def hash
-    to_a.hash
+    [@name, @version, @platform, @content_address, @ruby_abi].hash
```
Comment (jenshenny) [inline lib/rubygems/name_tuple.rb]: Should we change `to_a` instead so the fields are consistent (`to_a` has only `@name, @version, @platform`) ?

File: lib/rubygems/name_tuple.rb (hunk 6 above)
Comment (girachawda) [reply lib/rubygems/name_tuple.rb:17]: The removal of the guard makes sense, but then it changes the object shape for non-CA tuples. Functionally it makes no difference, but I kept the guard to avoid changing the shape of every existing tuple when only CA tuples need those fields. WDYT?

File: lib/rubygems/query_utils.rb [hunk 10]
```diff
@@ -254,19 +282,51 @@ def entry_versions(entry, name_tuples, platforms, specs)
...
+    end
+  end
+
+  def ruby_abi_metadata?(platform_ruby_abis)
+    platform_ruby_abis.values.any? do |ruby_abis_by_platform|
+      ruby_abis_by_platform.values.any?(&:any?)
+    end
+  end
+
+  def version_platform_labels(version, platforms, platform_ruby_abis, label_platform: false)
```
Comment (girachawda) [inline lib/rubygems/query_utils.rb:298]: this builds the platform/Ruby ABI text shown next to a version

File: lib/rubygems/query_utils.rb (hunk 7 above)
Comment (girachawda) [reply lib/rubygems/query_utils.rb]: That is correct! Changed it to just platform

File: lib/rubygems/name_tuple.rb [hunk 11]
```diff
@@ -46,12 +48,15 @@ def self.null
...
+    full_name = "#{@name}-#{@version}"
+    suffix = @content_address || platform_suffix
+    suffix ? "#{full_name}-#{suffix}" : full_name
+  end
+
+  private def platform_suffix # :nodoc:
+    return if @platform.nil? || @platform.empty? || @platform == Gem::Platform::RUBY
+
+    @platform
   end
```
Comment (jenshenny) [inline lib/rubygems/name_tuple.rb]: This is better but I don't think we need a separate method. This would match the shape of the other callsites like Specification.


```suggestion
    if @content_address
      "#{@name}-#{@version}-#{@content_address}"
    elsif @platform.nil? || @platform.empty? || @platform == Gem::Platform::RUBY
      "#{@name}-#{@version}"
    else
      "#{@name}-#{@version}-#{platform}"
    end
end
```

File: lib/rubygems/name_tuple.rb (hunk 6 above)
Comment (jenshenny) [reply lib/rubygems/name_tuple.rb:17]: Ah gotcha, we aren't setting the ivar unless it's set. Yeah, makes sense thanks!

File: lib/rubygems/query_utils.rb
Comment (jenshenny) [inline lib/rubygems/query_utils.rb:1]: This reads much better!! Though I think it can still be improved a bit more since there's a indirection to a lot of methods. I think the cause is that the formatting between all the different cases are drastically different (new line, no brackets on multiline). I like to propose this format which would have better consistency throughout the cases which I would believe would make the code a lot more simpler by creating more similarities and shared logic. 

lmk wdyt! I also am suggesting different naming for Ruby ABI since it's more human readable.

`a (1 x86_64-linux)`
```
a (2 platform: arm64-darwin, Ruby ABI: 3.5
   2 platform: x86_64-linux, Ruby ABI: 3.4
   1 platform: arm64-darwin, Ruby ABI: 3.4
   1 platform: x86_64-linux, Ruby ABI: 3.3, 3.4)
```
`a (2 Platform: arm64-darwin, Ruby ABI: 3.5)`

File: lib/rubygems/source.rb [hunk 12]
```diff
@@ -258,16 +258,26 @@ def load_compact_index_specs(type)
...
+
+        if Gem::ContentAddress.match?(suffix)
+          content_addressable_rows << [row_name, version, version_string, suffix]
+          next
+        end
+
+        Gem::NameTuple.new(row_name, version, suffix)
       end
 
+      gem_tuples.concat content_addressable_tuples(name, content_addressable_rows) unless content_addressable_rows.empty?
```
Comment (jenshenny) [inline lib/rubygems/source.rb]: I like how we're moving this to the end, but to my knowledge, this still fetches the gem info for every CA gem. Can we move this call to when we're finished filtering in query utils so we only fetch the gems we're returning?

File: lib/rubygems/query_utils.rb
Comment (girachawda) [reply lib/rubygems/query_utils.rb:1]: Made this change! So the methods that were split off for multiline vs inline formatting are gone. With the performance improvement of not looping through every NameTuple for CA gem info unless needed, the query_utils file now has to request source to complete the CA gem info.

File: lib/rubygems/query_utils.rb [hunk 13]
```diff
@@ -154,9 +154,37 @@ def show_remote_gems(name)
...
+    end
+  end
+
+  def latest_platform_key(tuple)
+    if tuple.content_address
+      [tuple.platform, tuple.ruby_abi || tuple.content_address]
+    else
+      tuple.platform
+    end
+  end
```
Comment (jenshenny) [inline lib/rubygems/query_utils.rb]: Could we call `source.max_versions_by_platform` instead of duplicating `latest_spec_tuples` and `latest_platform_key(tuple)`. I think they are copies of the `Source` methods?

Or could we get `source.decode_content_addressable_tuples` to call `max_versions_by_platform`

File: lib/rubygems/query_utils.rb [hunk 14]
```diff
@@ -195,90 +223,127 @@ def output_query_results(spec_tuples)
...
+
+      platforms.sort.reverse.flat_map do |version, pls|
+        version_label = version_label(version, specs)
+        labels = version_platform_labels(version, pls, platform_ruby_abis, label_platform: label_platform)
+        labels.empty? ? [version_label] : labels.map {|label| "#{version_label} #{label}" }
       end
+    end
 
-    entry << " (#{list.join ", "})"
+    separator = ruby_abi_metadata?(platform_ruby_abis) && list.length > 1 ? "\n#{" " * (entry.first.length + 2)}" : ", "
```
Comment (jenshenny) [inline lib/rubygems/query_utils.rb]: Or else `gem info c -r --all ` would look like this:

```
c (3
   2
   1)
    Platforms:
        1: x86_64-linux
        2: x86_64-linux
        3: arm64-darwin
```

```suggestion
    use_multiline_separator = !options[:details] && ruby_abi_metadata?(platform_ruby_abis) && list.length > 1
    separator = use_multiline_separator ? "\n#{" " * (entry.first.length + 2)}" : ", "
```

File: lib/rubygems/source.rb [hunk 15]
```diff
@@ -184,6 +184,31 @@ def compact_index_client # :nodoc:
...
+      [[tuple.name, tuple.version, tuple.content_address], tuple]
+    end
+
+    decoded = tuples.filter_map do |tuple|
+      next tuple unless tuple.content_address
+
+      decoded_by_key[[tuple.name, tuple.version, tuple.content_address]]
+    end
+
+    latest ? max_versions_by_platform(decoded) : decoded
```
Comment (jenshenny) [inline lib/rubygems/source.rb]: `max_versions_by_platform(decoded)` will return the max versions by platform for the all gems and not per gem. So this will return the latest gems per platform. We should call `max_versions_by_platform(decoded)` for each gem name.

Example Gems
aa 7.1 ruby, ab 1.5 ruby, ac 6.4 ruby, ad 1.16 x86_64-linux

Expected return for `latest`
all will be returned since they are all different gems

Actual
aa 7.1 ruby, ad 1.16 x86_64-linux is returned because the method will only return one gem per platform.

File: lib/rubygems/query_utils.rb [hunk 16]
```diff
@@ -229,44 +242,71 @@ def entry_details(entry, detail_tuple, specs, platforms)
...
-            out = platform_list.unshift(out).join(" ")
-          end
-
-          out
+        label_platform = ruby_abi_metadata?(platform_ruby_abis)
+
+        platforms.sort.reverse.flat_map do |version, pls|
+          out = version_label(version, specs)
+          labels = version_platform_labels(version, pls, platform_ruby_abis, label_platform: label_platform)
+          labels.empty? ? [out] : labels.map {|label| "#{out} #{label}" }
```
Comment (jenshenny) [inline lib/rubygems/query_utils.rb:267]: This currently outputs `nokogiri (1.16.0 arm64-darwin, 1.16.0 java, 1.16.0 x86_64-linux)` (duplicated versions). I believe this will cover the case for versions for fat and skinny binaries

```ruby
    list =
      if platforms.empty? || options[:details]
        name_tuples.map(&:version).uniq
      elsif ruby_abi_metadata?(platform_ruby_abis)
        platforms.sort.reverse.flat_map do |version, pls|
          out = version_label(version, specs)
          labels = version_platform_labels(version, pls, platform_ruby_abis, label_platform: true)
          next [out] if labels.empty?

          labels.map { |label| "#{out} #{label}" }
        end
      else
        platforms.sort.reverse.flat_map do |version, pls|
          out = version_label(version, specs)
          labels = version_platform_labels(version, pls, platform_ruby_abis, label_platform: false)
          next [out] if labels.empty?

          ["#{out} #{labels.join(" ")}"]
        end
      end
```

**Review by jenshenny — COMMENTED** [review]
A few small regressions with our refactor with the formatting, but otherwise lgtm

Would you also be able to scan the test coverage and add any edge cases that weren't tested for since tests should have caught these regressions?

File: lib/rubygems/query_utils.rb (hunk 16 above)
Comment (girachawda) [reply lib/rubygems/query_utils.rb:267]: Great catch! it now outputs: nokogiri (1.16.0 arm64-darwin java x86_64-linux)

File: test/rubygems/test_gem_safe_marshal.rb [hunk 17]
```diff
@@ -344,6 +344,41 @@ def test_rational
     end
   end
 
+  def test_name_tuple_unmarshall_content_addressable_metadata
```
Comment (jenshenny) [inline test/rubygems/test_gem_safe_marshal.rb]:
```suggestion
  def test_name_tuple_unmarshal_content_addressable_metadata
```

File: test/rubygems/test_gem_commands_info_command.rb
Comment (jenshenny) [inline test/rubygems/test_gem_commands_info_command.rb:1]: We should have a test case where there's multi CA / fat versions for a gem to see if it's formatted correctly (same with list and search).

**Review by jenshenny — APPROVED** [review]

File: test/rubygems/test_gem_commands_info_command.rb
Comment (girachawda) [reply test/rubygems/test_gem_commands_info_command.rb:1]: Added for list search and info!
